### PR TITLE
lsb-release: Fix/replace with a custom Bash script

### PR DESCRIPTION
**Pull request by primeos**
###### Motivation for this change

See #22729 and #23175.

TODOs:
- [x] Choose a license (I used MIT because Nixpkgs uses it, but we could also use something else)
- [x] Replace cat and getopt with references to the Nix store
- [x] Determine if we want to fallback to other files than `/etc/os-release`
  - Could be nice for non-NixOS systems, but probably problematic to implement as the format of the 
other files is most likely not standardized.
  - We'll skip this for now.
- [x] Verify the output (LGTM, but I might have missed something)
  - Thanks @davidak for checking it as well and posting additional examples
- [x] Verify that the script is secure and robust
  - Note: Some variables are unnecessary quoted (e.g. the ones which are either 0 or 1) and the variables sourced from `/etc/os-releases` are not quoted (but that file should always be owned by root).
  - Should be ok, all important variables are quoted. And this script doesn't run with any elevated privileges.
- ...

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/os-specific/linux/lsb-release/lsb_release.sh [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+      ;;
+  esac
+  shift
+done
+
+#  Read our variables.
+if [[ -e /etc/os-release ]]; then
+  . /etc/os-release
+else
+  echo "/etc/os-release is not present.  Aborting" >&2
```
Comment (davidak) [inline pkgs/os-specific/linux/lsb-release/lsb_release.sh]:
```suggestion
  echo "/etc/os-release is not present. Aborting" >&2
```
Or was it intentional?

File: pkgs/os-specific/linux/lsb-release/lsb_release.sh [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+if [[ "$all" = "1" ]] || [[ "$codename" = "1" ]]; then
+  if [[ "$short" = "0" ]]; then
+    printf "Codename:\t"
+  fi
+  echo $VERSION_CODENAME
+fi
+
+# Success
+exit 0
```
Comment (davidak) [inline pkgs/os-specific/linux/lsb-release/lsb_release.sh]: I think this is not needed since it is the default behaviour.

**Review by davidak — CHANGES_REQUESTED** [review]

**Comment by davidak** [comment]
Re: TODOS

>Choose a license (I used MIT because Nixpkgs uses it, but we could also use something else)

For inclusion in this repo MIT is best.

>Determine if we want to fallback to other files than /etc/os-release

Can be done later. Should not delay this PR.

Also, just support the standard lines.

>Verify the output (LGTM, but I might have missed something)

This script has an additional newline at the end of the output.

Except that identical to debian and ubuntu. :+1: 

```
[davidak@nixos:~]$ /nix/store/slm41d7vfqc6i429d0hn5341li3pi0m6-lsb_release/bin/lsb_release -a
No LSB modules are available.
Distributor ID:	NixOS
Description:	NixOS 19.03.172866.4649b6ef4b5 (Koi)
Release:	19.03.172866.4649b6ef4b5
Codename:	koi

```

```
root@d646ab96b21c:/# lsb_release -a
No LSB modules are available.
Distributor ID:	Debian
Description:	Debian GNU/Linux 9.5 (stretch)
Release:	9.5
Codename:	stretch
```

```
root@ab88485573f0:/# lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.2 LTS
Release:	18.04
Codename:	bionic
```

File: pkgs/os-specific/linux/lsb-release/lsb_release.sh (hunk 1 above)
Comment (primeos) [reply pkgs/os-specific/linux/lsb-release/lsb_release.sh]: Don't think so - I also noticed it, but I ignored if for now, but two spaces does seem strange... :)
(Except for Emacs users, IIRC)

File: pkgs/os-specific/linux/lsb-release/lsb_release.sh (hunk 2 above)
Comment (primeos) [reply pkgs/os-specific/linux/lsb-release/lsb_release.sh]: Right, forgot to remove that from the original Gist (probably made sense in some other context).

**Comment by primeos** [comment]
> This script has an additional newline at the end of the output.

That also tricked me but this is due to the configuration of `nix-shell`, which prints an additional newline after each command (if you pipe to vim, xxd, etc. there's only one newline like on Debian, etc.).
Not sure where this is configured.

But I just noticed that the script from Debian prints `No LSB modules are available.` to stderr, I'll change that in the next commit.

Thanks for pasting the outputs on other operating systems, that's really helpful :) (I only compared it to the output on Debian stretch)

**Comment by davidak** [comment]
I think the newline is no problem. It's only different.

>Thanks for pasting the outputs on other operating systems, that's really helpful :) (I only compared it to the output on Debian stretch)

I just used docker to quickly spin up an instance. Perfectly for such tests :)

**Comment by primeos** [comment]
> I think the newline is no problem. It's only different.

What I meant is that this is caused by the shell, not by the program (`lsb_release`):

```
$ lsb_release -a
No LSB modules are available.
Distributor ID:	NixOS
Description:	NixOS 19.03.git.50e5e3a (Koi)
Release:	19.03.git.50e5e3a
Codename:	koi

$ PS1='\[\033[1;32m\][nix-shell:\w]\$\[\033[0m\] '
$ lsb_release -a
No LSB modules are available.
Distributor ID:	NixOS
Description:	NixOS 19.03.git.50e5e3a (Koi)
Release:	19.03.git.50e5e3a
Codename:	koi
$ 
```

The prompt is configured to print an additional newline (via `PS1`, I did initially miss that, when I had a look).

So the output should actually be identical now (apart from the actual values of course).

**Review by davidak — APPROVED** [review]
Code looks sane and script works as expected. Thanks!

**Comment by davidak** [comment]
@primeos is this ready for merging or do you want to change anything before?

**Comment by primeos** [comment]
> @primeos is this ready for merging or do you want to change anything before?

@davidak  Sorry for the delay, I ran out of time (the time of this PR was a bit suboptimal).
I just checked the TODOs and it should be ready to go :)

cc @dtzWill @rardiol (previous contributors to `lsb-release`): Any objections? If not I would merge this PR in a few days.

**Comment by primeos** [comment]
> If not I would merge this PR in a few days.

Unfortunately I was pretty busy and lost track of this PR...
I just ran the rebuilds and unfortunately this broke the build for `radiotray-ng`.
Turns out that the output wasn't backwards compatible if `/etc/os-release` is missing (e.g. in the Nix build sandbox).

I did try my best to make the output as backwards compatible as possible (at least without `/etc/os-release` as the current binary doesn't work correctly outside of the sandbox and is therefore a bad example).
If the rebuilds succeed I'll merge this PR - let's hope there won't be any problematic regressions... (but in that case we should still be able to modify the script accordingly).

Unfortunately the code isn't really nice anymore but I guess it'll have to do for now... (we can still refactor it later if it works as intended). Also: We might want to change the default values if we're in a Nix build sandbox to increase the usefulness of the new script (currently it only helps at run-time).

**Pull request merged**